elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,281 @@
+
+import logging
+from itertools import combinations
+
+import numpy as np
+from scipy.spatial import cKDTree
+from scipy.special import digamma, gamma
+
+import elfi
+
+logger = logging.getLogger(__name__)
+
+
+class TwoStageSelection:
+    """Perform the summary-statistics selection proposed by Nunes and Balding (2010).
+
+    The rationale of the procedure is the following:
+    - Initially, the combinations of the to-be-assessed summary statistics is computed;
+    - In Stage 1, each summary-statistics combination is evaluated using the
+    minimum-entropy algorithm;
+    - In Stage 2, the minimum-entropy combination is selected,
+    and the `closest' datasets are identified;
+    - Further in Stage 2, for each summary-statistics combination,
+    the mean root sum of squared errors (MRSSE) is calculated over all `closest datasets',
+    and the minimum-MRSSE combination is chosen as the one with the optimal performance.
+
+    Notes
+    -----
+    Future extension: include different distance functions in the summary-statistics combinations.
+
+    Attributes
+    ----------
+    list_ss : array_like
+        List of the to-be-assessed summary statistics.
+    simulator : elfi.Simulator
+        Simulator initialised with the observations.
+    fn_distance : str or function
+        Function for calculating distance (can be a string if the function is implemented in ELFI).
+    seed : int, optional
+
+    References
+    ----------
+    [1] Nunes, M. A., & Balding, D. J. (2010).
+    On optimal selection of summary statistics for approximate Bayesian computation.
+    Statistical applications in genetics and molecular biology, 9(1).
+    [2] Blum, M. G., Nunes, M. A., Prangle, D., & Sisson, S. A. (2013).
+    A comparative review of dimension reduction methods in approximate Bayesian computation.
+    Statistical Science, 28(2), 189-208.
+
+    """
+
+    def __init__(self, list_ss, simulator, fn_distance,
+                 seed=0, max_cardinality=4, chosen_candidates=None):
+        """Initialise the summary-statistics selection based on the Two-Stage Procedure.
+
+        Parameters
+        ----------
+        list_ss : array_like
+            List of the to-be-assessed summary statistics.
+        simulator : elfi.Simulator
+            Simulator initialised with the observations.
+        fn_distance : str or function
+            Function for calculating distance
+            (can be a string if the function is implemented in ELFI).
+        seed : int, optional
+        max_cardinality : int
+            Maximum cardinality of a candidate summary-statistics sub-set.
+        chosen_candidates : array_like, optional
+            Set of candidate summary-statistics combinations provided by the user.
+
+        """
+        self.simulator = simulator
+        self.fn_distance = fn_distance
+        self.seed = seed
+        self.ss_candidates = self._init_ss_candidates(list_ss, max_cardinality=max_cardinality,
```

Comment:
If the user gives a list of combinations, no need to process it further. (Except maybe check validity, but then again we have few checks for such anyway.)